src/components/shapes/draw.js

```diff
@@ -579,6 +600,249 @@ function setupDragElement(gd, shapePath, shapeOptions, index, shapeLayer, editHe
+        var y2p = helpers.getDataToPixel(gd, ya, true, yRefType);
+        shapex0 = x2p(options.x0);
+        shapex1 = x2p(options.x1);
+        shapey0 = y2p(options.y0);
+        shapey1 = y2p(options.y1);
+    }
+
+    // Handle `auto` angle
+    var textangle = options.label.textangle;
+    if(textangle === 'auto') {
+        if(options.type === 'line') {
+            // Auto angle for line is same angle as line
+            textangle = calcTextAngle(shapex0, shapey0, shapex1, shapey1);
+        } else {
+            // Auto angle for all other shapes is 0
+            textangle = 0;
+        }
+    }
+
+    // Do a fake render so we can get the text bounding box height
+    var _labelText = labelGroup.append('text')
+        .attr(labelTextAttrs)
+        .classed('shape-label-text', true)
+        .text(text);
+    _labelText.call(function(s) {
+        s.call(Drawing.font, font).attr({});
+        svgTextUtils.convertToTspans(s, gd);
+        return s;
+    });
+    var textBB = Drawing.bBox(_labelText.node());
+    _labelText.remove();
+
+    // Calculate correct (x,y) for text
+    // We also determine true xanchor since xanchor depends on position when set to 'auto'
+    var textPos = calcTextPosition(shapex0, shapey0, shapex1, shapey1, options, textangle, textBB);
+    var textx = textPos.textx;
+    var texty = textPos.texty;
+    var xanchor = textPos.xanchor;
+
+    function textLayout(s) {
+        s.call(Drawing.font, font)
+            .attr({
+                'text-anchor': {
+                    left: 'start',
+                    center: 'middle',
+                    right: 'end'
+                }[xanchor],
+                'y': texty,
+                'x': textx,
+                'transform': 'rotate(' + textangle + ',' + textx + ',' + texty + ')'
+            });
+        svgTextUtils.convertToTspans(s, gd);
+        return s;
+    }
+
+    labelText.call(textLayout);
+}
+
+function calcTextAngle(shapex0, shapey0, shapex1, shapey1) {
+    var dy, dx;
+    dx = Math.abs(shapex1 - shapex0);
+    if(shapex1 >= shapex0) {
+        dy = shapey0 - shapey1;
+    } else {
+        dy = shapey1 - shapey0;
+    }
+    return -180 / Math.PI * Math.atan2(dy, dx);
+}
+
+function calcTextPosition(shapex0, shapey0, shapex1, shapey1, shapeOptions, actualTextAngle, textBB) {
+    var textPosition = shapeOptions.label.textposition;
+    var textAngle = shapeOptions.label.textangle;
+    var textPadding = shapeOptions.label.padding;
+    var shapeType = shapeOptions.type;
+    var textAngleRad = Math.PI / 180 * actualTextAngle;
+    var xanchor = shapeOptions.label.xanchor;
+    var yanchor = shapeOptions.label.yanchor;
+
+    var textx, texty, paddingX, paddingY;
+
+    // Text position functions differently for lines vs. other shapes
+    if(shapeType === 'line') {
+        // Set base position for start vs. center vs. end of line (default is 'center')
+        if(textPosition === 'start') {
```

Comment:
Wondering if you could possibly cast these line specific `textpositions` to regular values so that we could reuse similar `textposition` block for other shape types?

Something like: 
```js
if(shapeType === 'line') textPosition += 'bottom';
```

Reply:
I can't think of any way of doing this that would simplify the code, because even if we cast `textposition` to regular values, the positioning logic is still different for lines vs. other shapes. Definitely open to it if you have a specific suggestion though.

Reply:
That's OK. Resolving.